### PR TITLE
spotify-player: add module

### PR DESCRIPTION
**Pull request by diniamo**
### Description

Adds a module for `spotify-player` with the fields `enable`, `package`, `settings`, `themes`, `keymaps`.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

### CODE REVIEW

**Comment by diniamo** [comment]
Do the commits really have to be renamed? Can't the pull request be merged while squashing?

If so, what to and how do I rename them?

**Comment by diniamo** [comment]
Should be good now, I think

File: modules/programs/spotify-player.nix [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+  options.programs.spotify-player = {
+    enable = mkEnableOption "spotify-player";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.spotify-player;
+      defaultText = literalExpression "pkgs.spotify-player";
+      description = "The package to use for the spotify-player binary.";
+    };
```
Comment (sumnerevans) [inline modules/programs/spotify-player.nix:18]:
```suggestion
    package = mkPackageOption pkgs "spotify-player" { };
```

File: modules/programs/spotify-player.nix [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        {file}`$XDG_CONFIG_HOME/spotify-player/app.toml`.
+
+        See
+        <https://github.com/aome510/spotify-player/blob/master/docs/config.md#general>
+        for the full list of options.
+      '';
+    };
+
+    themes = mkOption {
+      type = types.listOf types.attrs;
```
Comment (sumnerevans) [inline modules/programs/spotify-player.nix:48]: not sure if this works, but I think it's better than just `types.attrs` if it works

```suggestion
      type = types.listOf tomlFormat.type;
```

File: modules/programs/spotify-player.nix [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        {file}`$XDG_CONFIG_HOME/spotify-player/theme.toml`.
+
+        See
+        <https://github.com/aome510/spotify-player/blob/master/docs/config.md#themes>
+        for the full list of options.
+      '';
+    };
+
+    keymaps = mkOption {
+      type = types.listOf types.attrs;
```
Comment (sumnerevans) [inline modules/programs/spotify-player.nix:99]:
```suggestion
      type = types.listOf tomlFormat.type;
```

File: tests/modules/programs/spotify-player/settings.nix [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        key_sequence = "M-enter";
+      }
+      {
+        command = "None";
+        key_sequence = "q";
+      }
+    ];
+  };
+
+  test.stubs.yazi = { };
```
Comment (sumnerevans) [inline tests/modules/programs/spotify-player/settings.nix:92]: I assume this was a copy-paste error.

```suggestion
  test.stubs.spotify-player = { };
```

**Review by sumnerevans — CHANGES_REQUESTED** [review]

**Comment by diniamo** [comment]
Suggestions, implemented, tests still run.

That last one is something else though.

**Review by sumnerevans — CHANGES_REQUESTED** [review]
Try rebasing on `master`? The tests are passing there.

Also, please squash this down to a single commit.

**Comment by diniamo** [comment]
If you check the error, then it's saying it can't find me in the maintainers list. I have no idea why though.

File: modules/programs/spotify-player.nix [hunk 5]
```diff
@@ -0,0 +1,150 @@
+{ config, lib, pkgs, ... }:
+let
+  inherit (lib)
+    mkEnableOption mkPackageOption mkOption types literalExpression mkIf;
+
+  cfg = config.programs.spotify-player;
+  tomlFormat = pkgs.formats.toml { };
+in {
+  meta.maintainers = with lib.maintainers; [ diniamo ];
```
Comment (sumnerevans) [inline modules/programs/spotify-player.nix:9]:
```suggestion
  meta.maintainers = with lib.hm.maintainers; [ diniamo ];
```

**Review by sumnerevans — CHANGES_REQUESTED** [review]
Ah, the issue is that `lib.maintainers` is the list of Nixpkgs maintainers, not the Home Manager maintainers.

**Comment by diniamo** [comment]
All done!

**Review by sumnerevans — APPROVED** [review]

**Comment by diniamo** [comment]
:sob:

**Comment by rycee** [comment]
Thanks! Merged to master now 🙂

**Comment by repparw** [comment]
Hi! i'm pretty new to nix and hm so I don't know what the proper way to raise this question is so I'm dropping it here, sorry!
Is there a proposal for adding actions to the module? It's a special type of keymap added to spotify-player in 0.19.1

`[[keymaps]]
command = "LyricPage"
key_sequence = "g l"

[[actions]]
action = "CopyLink"
key_sequence = "y"
target = "PlayingTrack"
`

it probably would be very similar to keymap adding an optional target (and it would have to be merged to the same keymap.toml file)

**Comment by diniamo** [comment]
How do actions differ from keymaps? I don't get it exactly.

**Comment by diniamo** [comment]
Although it seems like we will be adding an `actions` option to avoid breaking changes.

**Comment by repparw** [comment]
apart from having a different heading in the file I think not much. I mean,
the commands are different, but I don't think the module checks for
validity on those?

**Pull request merged**